test/integration/cluster_upstream_extension_integration_test.cc

```diff
@@ -0,0 +1,97 @@
+#include "envoy/config/bootstrap/v3/bootstrap.pb.h"
+#include "envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.pb.h"
+#include "envoy/registry/registry.h"
+#include "envoy/router/router.h"
+
+#include "common/buffer/buffer_impl.h"
+
+#include "test/integration/http_integration.h"
+#include "test/integration/upstreams/per_host_upstream_config.h"
+#include "test/test_common/registry.h"
+
+#include "fake_upstream.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+namespace {
+class ClusterUpstreamExtensionIntegrationTest
+    : public testing::TestWithParam<Network::Address::IpVersion>,
+      public HttpIntegrationTest {
+public:
+  ClusterUpstreamExtensionIntegrationTest()
+      : HttpIntegrationTest(Http::CodecClient::Type::HTTP1, GetParam()) {}
+
+  void SetUp() override {
+    setDownstreamProtocol(Http::CodecClient::Type::HTTP1);
+    setUpstreamProtocol(FakeHttpConnection::Type::HTTP1);
+  }
+
+  void populateMetadataTestData(envoy::config::core::v3::Metadata& metadata, const std::string& k1,
+                                const std::string& k2, const std::string& value) {
+
+    ProtobufWkt::Struct struct_obj;
+    (*struct_obj.mutable_fields())[k2] = ValueUtil::stringValue(value);
+
+    (*metadata.mutable_filter_metadata())[k1] = struct_obj;
+  }
+
+  void initialize() override {
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      auto* cluster = bootstrap.mutable_static_resources()->mutable_clusters(0);
+      cluster->mutable_upstream_config()->set_name("envoy.filters.connection_pools.http.per_host");
+      cluster->mutable_upstream_config()->mutable_typed_config();
+      populateMetadataTestData(*cluster->mutable_metadata(), "foo", "bar", "cluster-value");
+      populateMetadataTestData(*cluster->mutable_load_assignment()
+                                    ->mutable_endpoints(0)
+                                    ->mutable_lb_endpoints(0)
+                                    ->mutable_metadata(),
+                               "foo", "bar", "host-value");
+    });
+    HttpIntegrationTest::initialize();
+  }
+  PerHostGenericConnPoolFactory per_host_upstream_factory_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ClusterUpstreamExtensionIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+TEST_P(ClusterUpstreamExtensionIntegrationTest,
+       VerifyRequestHeadersAreRewrittenByClusterAndHostMetadata) {
```

Comment:
Could you add some comments here about what we're trying to verify here? Just to make it easier for future readers to grok what this test is about.

Reply:
add comment